SnoopCompileCore/src/snoopi.jl

```diff
@@ -86,6 +86,74 @@ function _snoopi(cmd::Expr, tmin = 0.0)
     end
 end
 
+function start_deep_timing()
+    Core.Compiler.Timings.reset_timings()
+    Core.Compiler.__set_measure_typeinf(true)
+end
+function stop_deep_timing()
+    Core.Compiler.__set_measure_typeinf(false)
+    Core.Compiler.Timings.close_current_timer()
+end
+
+function finish_snoopi_deep()
+    return Core.Compiler.Timings._timings[1]
+end
+
+function _snoopi_deep(cmd::Expr)
+    return quote
+        start_deep_timing()
+        try
+            $(esc(cmd))
+        finally
+            stop_deep_timing()
+        end
+        finish_snoopi_deep()
+    end
+end
+
+"""
+    timing_tree = @snoopi_deep commands
+
+Produce a profile of julia's type inference, containing the amount of time spent infering
+for every `MethodInstance` processed while executing `commands`.
+
+The top-level node in this profile tree is `ROOT`, which contains the time spent _not_ in
+julia's type inference (codegen, llvm_opt, runtime, etc).
+
+To make use of these results, see the processing functions in SnoopCompile:
+    - [`SnoopCompile.flatten_times(timing_tree)`](@ref)
+    - [`SnoopCompile.to_flamegraph(timing_tree)`](@ref)
+
+# Examples
+```julia
+julia> timing = SnoopCompileCore.@snoopi_deep begin
+           @eval sort(rand(100))  # Evaluate some code and profile julia's type inference
+       end;
+
+julia> using SnoopCompile, ProfileView
+
+julia> times = SnoopCompile.flatten_times(timing, tmin_secs=0.001)
+4-element Vector{Any}:
+ 0.001088448 => Core.Compiler.Timings.InferenceFrameInfo(MethodInstance for fpsort!(::Vector{Float64}, ::Base.Sort.QuickSortAlg, ::Base.Order.ForwardOrdering), 0x00000000000072e4, Any[], Any[Vector{Float64}, Core.Const(Base.Sort.QuickSortAlg()), Core.Const(Base.Order.ForwardOrdering()), Union{}, Union{}, Union{}, Union{}, Union{}, Union{}, Union{}, Union{}, Union{}])
+ 0.001618478 => Core.Compiler.Timings.InferenceFrameInfo(MethodInstance for rand!(::Random.MersenneTwister, ::Random.UnsafeView{Float64}, ::Random.SamplerTrivial{Random.CloseOpen01{Float64}, Float64}), 0x00000000000072e4, Any[], Any[Random.MersenneTwister, Random.UnsafeView{Float64}, Core.Const(Random.SamplerTrivial{Random.CloseOpen01{Float64}, Float64}(Random.CloseOpen01{Float64}())), Union{}, Union{}, Union{}, Union{}, Union{}, Union{}, Union{}])
+ 0.002289655 => Core.Compiler.Timings.InferenceFrameInfo(MethodInstance for _rand_max383!(::Random.MersenneTwister, ::Random.UnsafeView{Float64}, ::Random.CloseOpen01{Float64}), 0x00000000000072e4, Any[], Any[Random.MersenneTwister, Random.UnsafeView{Float64}, Core.Const(Random.CloseOpen01{Float64}()), Union{}, Union{}, Union{}, Union{}])
+ 0.093143594 => Core.Compiler.Timings.InferenceFrameInfo(MethodInstance for ROOT(), 0x0000000000000000, Any[], Any[])
```

Comment:
Haha woah, these are really long. Sorry, i didn't realize because my editor doesn't wrap.

Is this example useful? Should I trim the output somehow?

Reply:
Probably best. Maybe make it `InferenceFrameInfo(MethodInstance for fpsort!...)`?

We might also want to add a docstring for `InferenceFrameInfo`.